test/sqllogictest/types.slt

```diff
@@ -596,9 +596,79 @@ SELECT '1'::materialize.pg_catalog.int4
 1
 
 # tables are not types yet
-query error "pg_catalog.pg_enum" does not refer to a type
+query error type "pg_catalog.pg_enum" does not exist
 SELECT '1'::pg_catalog.pg_enum
 
+# relations can have the same name as built-in types
+statement ok
+CREATE VIEW int4 AS VALUES (1)
+
+query I
+SELECT * FROM int4
+----
+1
+
+# but within the same schema, types cannot have the same name as a relation
+statement error view "materialize.public.int4" already exists
+CREATE TYPE int4 AS (a int)
+
+# creating relations with the same name as an existing type is not allowed...
+
+statement ok
+CREATE TYPE rectype AS (a int)
+
+statement error type "materialize.public.rectype" already exists
+CREATE VIEW rectype AS VALUES (1)
+
+statement error type "materialize.public.rectype" already exists
+CREATE MATERIALIZED VIEW rectype AS VALUES (1)
+
+statement error type "materialize.public.rectype" already exists
+CREATE TABLE rectype (a int)
+
+statement error type "materialize.public.rectype" already exists
+CREATE SOURCE rectype FROM LOAD GENERATOR COUNTER
+
+# ...not even via rename...
+
+statement ok
+CREATE VIEW rectype_sneaky_v AS VALUES (1)
+
+statement error catalog item 'rectype' already exists
+ALTER VIEW rectype_sneaky_v RENAME TO rectype
+
+statement ok
+CREATE MATERIALIZED VIEW rectype_sneaky_mv AS VALUES (1)
+
+statement error catalog item 'rectype' already exists
+ALTER MATERIALIZED VIEW rectype_sneaky_mv RENAME TO rectype
+
+statement ok
+CREATE TABLE rectype_sneaky_t (a int)
+
+statement error catalog item 'rectype' already exists
+ALTER TABLE rectype_sneaky_t RENAME TO rectype
+
+statement ok
+CREATE SOURCE rectype_sneaky_s FROM LOAD GENERATOR COUNTER
+
+statement error catalog item 'rectype' already exists
+ALTER SOURCE rectype_sneaky_s RENAME TO rectype
+
+# creating indexes with the same name as a type is ok though...
+
+statement ok
+CREATE INDEX rectype ON text_to_regproc (a)
+
+statement ok
+DROP INDEX rectype
+
+statement ok
+CREATE INDEX rectype_sneaky ON text_to_regproc (a)
+
+statement ok
+ALTER INDEX rectype_sneaky RENAME TO rectype
```

Comment:
Are you sure about this? I see something different in PostgreSQL:

```
postgres=# CREATE SCHEMA s;
CREATE SCHEMA
postgres=# SET search_path TO s;
SET
postgres=# CREATE TABLE t (a INT);
CREATE TABLE
postgres=# CREATE TYPE i AS (a INT);
CREATE TYPE
postgres=# CREATE INDEX i ON t(a);
ERROR:  relation "i" already exists

```

Reply:
Ugh, good point, it's just non-record types that are allowed to coexist in PostgreSQL. I've disabled mixing indexes and _all_ types of types in Materialize for now, to be maximally restrictive. We can weaken in the future if we like.